File: zilli/training/rl_trainer.py
from typing import Dict, List, Optional
from zilli.training.cispo import CISPO_Trainer
from zilli.training.grpo import GRPO_Trainer


class RLTrainer:
    def __init__(self, config: Dict):
        self.config = config
        algorithm = config.get("algorithm", "CISPO").upper()
        if algorithm == "CISPO":
            self.impl = CISPO_Trainer(config)
        elif algorithm == "GRPO":
            self.impl = GRPO_Trainer(config)
        else:
            raise ValueError(f"Unknown algorithm: {algorithm}")

    def update(self, batch: List[Dict]) -> Dict[str, float]:
        if self.config.get("algorithm", "").upper() == "GRPO":
            advantages = self.impl.compute_advantages(batch)
        else:
            rewards = [t.get("reward", 0.0) for t in batch]
            dones = [t.get("done", False) for t in batch]
            advantages = self.impl.compute_advantages(rewards, dones)

        return self.impl.compute_loss(batch, advantages)

File: zilli-rs/src/training/rl_trainer.rs
use super::cispo::CispoTrainer;
use super::config::TrainingConfig;
use super::grpo::GrpoTrainer;

pub enum SelectedTrainer {
    CISPO(CispoTrainer),
    GRPO(GrpoTrainer),
}

pub struct RLTrainer {
    trainer: SelectedTrainer,
}

impl RLTrainer {
    pub fn new(config: Option<TrainingConfig>) -> Self {
        let config = config.unwrap_or_default();
        let trainer = match config.algorithm.as_str() {
            "grpo" => SelectedTrainer::GRPO(GrpoTrainer::new(config)),
            _ => SelectedTrainer::CISPO(CispoTrainer::new(config)),
        };
        Self { trainer }
    }

    pub fn update(&self, batch: &[f64]) -> serde_json::Value {
        let dones = vec![false; batch.len()];
        let group_batch = vec![batch.to_vec()];

        let advantages = match &self.trainer {
            SelectedTrainer::CISPO(t) => t.compute_advantages(batch, &dones),
            SelectedTrainer::GRPO(t) => t.compute_advantages(&group_batch),
        };

        match &self.trainer {
            SelectedTrainer::CISPO(t) => t.compute_loss(&advantages),
            SelectedTrainer::GRPO(t) => t.compute_loss(batch, &advantages),
        }
    }
}
